Deprecate / refactor EventDispatcher's multitude of send/receive/yield/handle events methods

gavin-norman-sociomantic opened the pull request.

Commit:
Deprecate EventDispatcher combinatorial send/receive/yield/events methods

Part of #128.

gavin-norman-sociomantic commented:
Blocked. Will test with dhtproto.

gavin-norman-sociomantic commented:
>Will test with dhtproto.

Interesting. This works without any deprecation messages. dhtproto must only use the simple send/receive methods and the `RequestEventDispatcher` methods (which aren't changed by this PR).

gavin-norman-sociomantic commented:
Tested with dmqproto. Compiles and tests with deprecation warnings, as expected. I tried converting a couple of `sendReceive` calls to use `nextEvent` and it seemed pretty straightforward. I feel this is the right road to go down.

Unblocked.

david-eckardt-sociomantic commented on src/swarm/neo/connection/RequestOnConnBase.d:
```diff
@@ -1014,22 +1014,13 @@ abstract class RequestOnConnBase
+            NextEventFlags flags;
+            flags |= NextEventFlags.Receive;
+            auto event = this.nextEvent(flags);
+            assert(event.active == event.active.received);
```
- Why not `auto event = this.nextEvent(NextEventFlags.Receive);`?
- The assertion is done already when evaluating `event.received`.

gavin-norman-sociomantic replied:
>Why not auto event = this.nextEvent(NextEventFlags.Receive);?

Right :man_facepalming: 

>The assertion is done already when evaluating event.received.

Ah, indeed.

gavin-norman-sociomantic replied:
Updated.

Commit:
Refactor EventDispatcher send/receive methods to use nextEvents internally

It seems reasonable to leave these very basic methods as a convenient
shorthand for common usages of nextEvent.

Part of #128.

Commit:
Add EventDispatcher.NextEventFlags.None

Useful when calling nextEvent without needing to receive or yield.

gavin-norman-sociomantic commented:
Rebased.

gavin-norman-sociomantic commented:
Auto-merged after David's approval.
